### PR TITLE
Add data source cyral_sidecar_listener

### PR DESCRIPTION
**Pull request by wcmjunior**
## Description of the change

Add a new data source `cyral_sidecar_listener` to allow searching for listeners from a given `sidecar_id` and filtering by repo type.

It is important to note that I took the conscious decision to reuse the bulk of the schema from the resource `cyral_sidecar_listener` for the return list in this new data source. The only difference from the JSON returned by the actual `sidecars/{sidecarId}/listeners` API is that the `sidecarId` is not returned as it is already part of the query being performed. I decided to use the `sidecar_id` provided by the user (it is a required parameter) to fill in this gap as shown [here](https://github.com/cyralinc/terraform-provider-cyral/blob/58cdcaf893508eecb5fbe5f600675e66f5c6ec53/cyral/data_source_cyral_sidecar_listener.go#L33).

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] Jira issue referenced in commit message and/or PR title

### Testing

All acceptance tests are passing as follows:

```
wilson@Wilsons-MacBook-Pro terraform-provider-cyral % make local/test
go test github.com/cyralinc/terraform-provider-cyral/... -v -race -timeout 20m
?       github.com/cyralinc/terraform-provider-cyral    [no test files]
=== RUN   TestNewClient_WhenTLSSkipVerifyIsEnabled_ThenInsecureSkipVerifyIsTrue
2023/06/16 00:41:42 [DEBUG] Init NewClient
2023/06/16 00:41:42 [DEBUG] TokenSource: &{0xc00013cd98 {0 0} <nil>}
2023/06/16 00:41:42 [DEBUG] End NewClient
--- PASS: TestNewClient_WhenTLSSkipVerifyIsEnabled_ThenInsecureSkipVerifyIsTrue (0.00s)
=== RUN   TestNewClient_WhenTLSSkipVerifyIsDisabled_ThenInsecureSkipVerifyIsFalse
2023/06/16 00:41:42 [DEBUG] Init NewClient
2023/06/16 00:41:42 [DEBUG] TokenSource: &{0xc00013cde0 {0 0} <nil>}
2023/06/16 00:41:42 [DEBUG] End NewClient
--- PASS: TestNewClient_WhenTLSSkipVerifyIsDisabled_ThenInsecureSkipVerifyIsFalse (0.00s)
=== RUN   TestNewClient_WhenClientIDIsEmpty_ThenThrowError
2023/06/16 00:41:42 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenClientIDIsEmpty_ThenThrowError (0.00s)
=== RUN   TestNewClient_WhenClientSecretIsEmpty_ThenThrowError
2023/06/16 00:41:42 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenClientSecretIsEmpty_ThenThrowError (0.00s)
=== RUN   TestNewClient_WhenControlPlaneIsEmpty_ThenThrowError
2023/06/16 00:41:42 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenControlPlaneIsEmpty_ThenThrowError (0.00s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/client     (cached)
=== RUN   TestAccDatalabelDataSource
=== PAUSE TestAccDatalabelDataSource
=== RUN   TestAccIntegrationIdPSAMLDataSource
=== PAUSE TestAccIntegrationIdPSAMLDataSource
=== RUN   TestAccLoggingIntegrationDataSource
=== PAUSE TestAccLoggingIntegrationDataSource
=== RUN   TestAccRepositoryDataSource
=== PAUSE TestAccRepositoryDataSource
=== RUN   TestAccRoleDataSource
=== PAUSE TestAccRoleDataSource
=== RUN   TestAccSAMLCertificateDataSource
=== PAUSE TestAccSAMLCertificateDataSource
=== RUN   TestAccSAMLConfigurationDataSource
=== PAUSE TestAccSAMLConfigurationDataSource
=== RUN   TestAccSidecarBoundPortsDataSource
=== PAUSE TestAccSidecarBoundPortsDataSource
=== RUN   TestAccSidecarCftTemplateDataSource
=== PAUSE TestAccSidecarCftTemplateDataSource
=== RUN   TestAccSidecarIDDataSource
=== PAUSE TestAccSidecarIDDataSource
=== RUN   TestAccSidecarInstanceIDsDataSource
=== PAUSE TestAccSidecarInstanceIDsDataSource
=== RUN   TestAccSidecarListenerDataSource
=== PAUSE TestAccSidecarListenerDataSource
=== RUN   TestIntegrationsData_GetValue_Default
--- PASS: TestIntegrationsData_GetValue_Default (0.00s)
=== RUN   TestIntegrationsData_GetValue_Splunk
--- PASS: TestIntegrationsData_GetValue_Splunk (0.00s)
=== RUN   TestAccProvider
2023/06/16 00:48:01 [DEBUG] Init dataSourceSidecarListener
2023/06/16 00:48:01 [DEBUG] End dataSourceSidecarListener
--- PASS: TestAccProvider (0.01s)
=== RUN   TestAccDatalabelResource
=== PAUSE TestAccDatalabelResource
=== RUN   TestAccDatadogIntegrationResource
=== PAUSE TestAccDatadogIntegrationResource
=== RUN   TestAccELKIntegrationResource
=== PAUSE TestAccELKIntegrationResource
=== RUN   TestAccHCVaultIntegrationResource
=== PAUSE TestAccHCVaultIntegrationResource
=== RUN   TestAccIntegrationIdPSAMLDraftResource
=== PAUSE TestAccIntegrationIdPSAMLDraftResource
=== RUN   TestAccIntegrationIdPSAMLResource
=== PAUSE TestAccIntegrationIdPSAMLResource
=== RUN   TestAccIdPIntegrationResource
=== PAUSE TestAccIdPIntegrationResource
=== RUN   TestAccLogsIntegrationResourceCloudWatch
=== PAUSE TestAccLogsIntegrationResourceCloudWatch
=== RUN   TestAccLogsIntegrationResourceDataDog
=== PAUSE TestAccLogsIntegrationResourceDataDog
=== RUN   TestAccLogsIntegrationResourceElk
=== PAUSE TestAccLogsIntegrationResourceElk
=== RUN   TestAccLogsIntegrationResourceSplunk
=== PAUSE TestAccLogsIntegrationResourceSplunk
=== RUN   TestAccLogsIntegrationResourceSumologic
=== PAUSE TestAccLogsIntegrationResourceSumologic
=== RUN   TestAccLogsIntegrationResourceFluentbit
=== PAUSE TestAccLogsIntegrationResourceFluentbit
=== RUN   TestAccLogstashIntegrationResource
=== PAUSE TestAccLogstashIntegrationResource
=== RUN   TestAccLookerIntegrationResource
=== PAUSE TestAccLookerIntegrationResource
=== RUN   TestAccDuoMFAIntegrationResource
=== PAUSE TestAccDuoMFAIntegrationResource
=== RUN   TestAccPagerDutyIntegrationResource
=== PAUSE TestAccPagerDutyIntegrationResource
=== RUN   TestAccSlackAlertsIntegrationResource
=== PAUSE TestAccSlackAlertsIntegrationResource
=== RUN   TestAccSplunkIntegrationResource
=== PAUSE TestAccSplunkIntegrationResource
=== RUN   TestAccSumoLogicIntegrationResource
=== PAUSE TestAccSumoLogicIntegrationResource
=== RUN   TestAccMsTeamsIntegrationResource
=== PAUSE TestAccMsTeamsIntegrationResource
=== RUN   TestAccPolicyRuleResource
=== PAUSE TestAccPolicyRuleResource
=== RUN   TestPolicyRuleResourceUpgradeV0
--- PASS: TestPolicyRuleResourceUpgradeV0 (0.00s)
=== RUN   TestAccPolicyResource
=== PAUSE TestAccPolicyResource
=== RUN   TestAccRepositoryAccessGatewayResource
=== PAUSE TestAccRepositoryAccessGatewayResource
=== RUN   TestAccRepositoryAccessRulesResource
=== PAUSE TestAccRepositoryAccessRulesResource
=== RUN   TestAccRepositoryBindingResource
=== PAUSE TestAccRepositoryBindingResource
=== RUN   TestAccRepositoryConfAnalysisResource
=== PAUSE TestAccRepositoryConfAnalysisResource
=== RUN   TestRepositoryConfAnalysisResourceUpgradeV0
--- PASS: TestRepositoryConfAnalysisResourceUpgradeV0 (0.00s)
=== RUN   TestAccRepositoryConfAuthResource
=== PAUSE TestAccRepositoryConfAuthResource
=== RUN   TestRepositoryConfAuthResourceUpgradeV0
--- PASS: TestRepositoryConfAuthResourceUpgradeV0 (0.00s)
=== RUN   TestAccRepositoryDatamapResource
=== PAUSE TestAccRepositoryDatamapResource
=== RUN   TestAccRepositoryNetworkAccessPolicyResource
=== PAUSE TestAccRepositoryNetworkAccessPolicyResource
=== RUN   TestAccRepositoryResource
=== PAUSE TestAccRepositoryResource
=== RUN   TestAccRepositoryUserAccountResource
=== PAUSE TestAccRepositoryUserAccountResource
=== RUN   TestAccRoleSSOGroupsResource
=== PAUSE TestAccRoleSSOGroupsResource
=== RUN   TestRoleSSOGroupsResourceUpgradeV0
--- PASS: TestRoleSSOGroupsResourceUpgradeV0 (0.00s)
=== RUN   TestAccRoleResource
=== PAUSE TestAccRoleResource
=== RUN   TestAccSidecarCredentialsResource
=== PAUSE TestAccSidecarCredentialsResource
=== RUN   TestSidecarListenerResource
=== PAUSE TestSidecarListenerResource
=== RUN   TestAccSidecarResource
=== PAUSE TestAccSidecarResource
=== RUN   TestElementsMatch
--- PASS: TestElementsMatch (0.00s)
=== CONT  TestAccIntegrationIdPSAMLDataSource
=== CONT  TestAccSidecarResource
=== CONT  TestAccRepositoryAccessRulesResource
=== CONT  TestAccLogsIntegrationResourceFluentbit
=== CONT  TestAccDatalabelResource
=== CONT  TestAccSidecarListenerDataSource
=== CONT  TestAccSidecarIDDataSource
=== CONT  TestAccSidecarInstanceIDsDataSource
=== CONT  TestAccSidecarCftTemplateDataSource
=== CONT  TestAccDatadogIntegrationResource
--- PASS: TestAccLogsIntegrationResourceFluentbit (6.26s)
=== CONT  TestAccSAMLConfigurationDataSource
--- PASS: TestAccDatadogIntegrationResource (6.71s)
=== CONT  TestAccSAMLCertificateDataSource
=== CONT  TestAccRoleDataSource
--- PASS: TestAccDatalabelResource (7.48s)
=== CONT  TestAccRepositoryDataSource
--- PASS: TestAccSidecarIDDataSource (9.00s)
--- PASS: TestAccSAMLCertificateDataSource (2.84s)
=== CONT  TestAccSlackAlertsIntegrationResource
--- PASS: TestAccSidecarCftTemplateDataSource (10.59s)
=== CONT  TestAccDatalabelDataSource
=== CONT  TestAccSplunkIntegrationResource
--- PASS: TestAccRepositoryAccessRulesResource (12.70s)
--- PASS: TestAccSidecarInstanceIDsDataSource (13.49s)
=== CONT  TestAccLogsIntegrationResourceSumologic
--- PASS: TestAccSAMLConfigurationDataSource (8.68s)
=== CONT  TestAccLogsIntegrationResourceElk
=== CONT  TestAccLogsIntegrationResourceDataDog
--- PASS: TestAccSlackAlertsIntegrationResource (6.09s)
=== CONT  TestAccLogsIntegrationResourceSplunk
--- PASS: TestAccRoleDataSource (9.04s)
--- PASS: TestAccSidecarListenerDataSource (18.59s)
=== CONT  TestAccHCVaultIntegrationResource
--- PASS: TestAccSplunkIntegrationResource (6.05s)
=== CONT  TestAccLogsIntegrationResourceCloudWatch
=== CONT  TestAccIntegrationIdPSAMLResource
--- PASS: TestAccLogsIntegrationResourceSumologic (6.34s)
=== CONT  TestAccMsTeamsIntegrationResource
--- PASS: TestAccIntegrationIdPSAMLDataSource (20.15s)
--- PASS: TestAccLogsIntegrationResourceDataDog (5.54s)
=== CONT  TestAccRepositoryAccessGatewayResource
--- PASS: TestAccLogsIntegrationResourceElk (7.07s)
=== CONT  TestAccRoleResource
=== CONT  TestAccELKIntegrationResource
--- PASS: TestAccLogsIntegrationResourceSplunk (5.90s)
=== CONT  TestAccRoleSSOGroupsResource
--- PASS: TestAccRepositoryDataSource (13.49s)
=== CONT  TestAccIntegrationIdPSAMLDraftResource
--- PASS: TestAccHCVaultIntegrationResource (5.97s)
--- PASS: TestAccLogsIntegrationResourceCloudWatch (6.16s)
=== CONT  TestAccDuoMFAIntegrationResource
--- PASS: TestAccMsTeamsIntegrationResource (5.31s)
=== CONT  TestAccPolicyResource
--- PASS: TestAccSidecarResource (25.87s)
=== CONT  TestAccRepositoryDatamapResource
=== CONT  TestAccPolicyRuleResource
--- PASS: TestAccDatalabelDataSource (17.45s)
--- PASS: TestAccELKIntegrationResource (7.09s)
=== CONT  TestAccRepositoryNetworkAccessPolicyResource
=== CONT  TestAccLoggingIntegrationDataSource
--- PASS: TestAccDuoMFAIntegrationResource (6.48s)
--- PASS: TestAccPolicyResource (6.66s)
=== CONT  TestAccPagerDutyIntegrationResource
=== CONT  TestAccSidecarBoundPortsDataSource
--- PASS: TestAccIntegrationIdPSAMLDraftResource (9.15s)
--- PASS: TestAccRoleSSOGroupsResource (14.28s)
=== CONT  TestAccIdPIntegrationResource
=== CONT  TestAccRepositoryConfAnalysisResource
--- PASS: TestAccPagerDutyIntegrationResource (6.57s)
--- PASS: TestAccIntegrationIdPSAMLResource (19.28s)
=== CONT  TestAccRepositoryResource
--- PASS: TestAccRepositoryAccessGatewayResource (18.23s)
=== CONT  TestAccSidecarCredentialsResource
--- PASS: TestAccLoggingIntegrationDataSource (9.07s)
=== CONT  TestAccSumoLogicIntegrationResource
=== CONT  TestSidecarListenerResource
--- PASS: TestAccRoleResource (20.69s)
--- PASS: TestAccSidecarCredentialsResource (5.46s)
=== CONT  TestAccRepositoryConfAuthResource
=== CONT  TestAccLogstashIntegrationResource
--- PASS: TestAccPolicyRuleResource (17.24s)
=== CONT  TestAccRepositoryUserAccountResource
--- PASS: TestAccSidecarBoundPortsDataSource (12.24s)
=== CONT  TestAccRepositoryBindingResource
--- PASS: TestAccRepositoryDatamapResource (22.34s)
=== CONT  TestAccLookerIntegrationResource
--- PASS: TestAccSumoLogicIntegrationResource (7.77s)
--- PASS: TestAccRepositoryConfAnalysisResource (9.79s)
--- PASS: TestAccRepositoryNetworkAccessPolicyResource (22.94s)
--- PASS: TestAccLookerIntegrationResource (7.54s)
--- PASS: TestAccLogstashIntegrationResource (12.56s)
--- PASS: TestAccRepositoryBindingResource (11.02s)
--- PASS: TestAccRepositoryConfAuthResource (15.78s)
--- PASS: TestAccRepositoryResource (22.06s)
--- PASS: TestAccIdPIntegrationResource (34.14s)
--- PASS: TestSidecarListenerResource (30.61s)
--- PASS: TestAccRepositoryUserAccountResource (41.01s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral      87.343s
```

### CODE REVIEW

**Review by VictorGFM — COMMENTED** [review]
Looking good, I have a few comments below for consideration.

File: cyral/data_source_cyral_sidecar_listener.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+	"github.com/google/uuid"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/cyralinc/terraform-provider-cyral/client"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	SidecarListenerListKey = "listener_list"
```
[the remark this answers is not in the document]
Comment (wcmjunior) [reply cyral/data_source_cyral_sidecar_listener.go:16]: We use `_list` to name the returned lists in most of our data sources.

File: cyral/data_source_cyral_sidecar_listener.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+			return fmt.Sprintf("https://%s/v1/sidecars/%s/listeners", c.ControlPlane, sidecarID)
+		},
+		NewResponseData: func(_ *schema.ResourceData) ResponseData { return &ReadDataSourceSidecarListenerAPIResponse{} },
+	}
+}
+
+func dataSourceSidecarListener() *schema.Resource {
+	log.Printf("[DEBUG] Init dataSourceSidecarListener")
+	listenerSchema := getSidecarListenerSchema()
+	for _, v := range listenerSchema {
```
[the remark this answers is not in the document]
Comment (wcmjunior) [reply cyral/data_source_cyral_sidecar_listener.go:78]: I did not know about this function. It makes more sense. Thanks!

File: cyral/data_source_cyral_sidecar_listener.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+			SidecarIDKey: {
+				Description: "Filter the results by sidecar ID.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			RepoTypesKey: {
+				Description: "Filter the results per repository type. Supported repo types:" + supportedTypesMarkdown(repositoryTypes()),
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
```
[the remark this answers is not in the document]
Comment (wcmjunior) [reply cyral/data_source_cyral_sidecar_listener.go]: I followed what we have in the `cyral_sidecar_listener` resource which is a list of strings, though this resource description states that `repo_types` should have a single value. I renamed it to singular so it makes more sense.

File: cyral/data_source_cyral_sidecar_listener.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (wcmjunior) [reply cyral/data_source_cyral_sidecar_listener.go]: Had to fix the function as it was not working properly. See [ea92ee3](https://github.com/cyralinc/terraform-provider-cyral/pull/410/commits/ea92ee3f82a3afbabba2e962fe3337c99413d8be).

**Review by VictorGFM — APPROVED** [review]
@wcmjunior Looks good!

**Pull request merged**